core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -1811,20 +1889,13 @@ object ZManaged {
                     res      <- managed.reserve
                     resource <- restore(res.acquire).onError(err => res.release(Exit.Failure(err)))
                     release  = res.release.andThen(_.provide(env))
-                    _        <- finalizers.update(_ + release)
-                    done <- Ref
-                             .make(release(Exit.Success(resource)).ensuring(finalizers.update(_ - release)))
-                             .map(_.modify((_, ZIO.unit)).flatten)
+                    _        <- finalizer.add(release)
+                    done     = finalizer.run(Exit.succeed(resource))
```

Comment:
I think this is slightly wrong; the previous formulation would only run the release for the current resource. This formulation runs *all* finalizers.

Reply:
Good point. Fixed now!